backend/src/main/java/com/yohoard/notes/TagService.java

```diff
@@ -0,0 +1,39 @@
+package com.yohoard.notes;
+
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Service
+public class TagService {
+
+    private final TagRepository tagRepository;
+
+    public TagService(TagRepository tagRepository) {
+        this.tagRepository = tagRepository;
+    }
+
+    public List<Tag> getOrCreate(List<String> names) {
+        List<Tag> existingTags = tagRepository.findByNameIn(names);
+
+        Set<String> existingNames = existingTags.stream()
+                .map(Tag::getName)
+                .collect(Collectors.toSet());
+
+        List<Tag> newTags = names.stream()
+                .filter(name -> !existingNames.contains(name))
+                .map(name -> new Tag(name, "", 0))
+                .toList();
+
+        tagRepository.saveAll(newTags);
+
+        List<Tag> allTags = new ArrayList<>(existingTags);
+
+        allTags.addAll(newTags);
```

Comment:
Przy tym całym marudzeniu, że set byłby lepszy niż lista 😄 myślę, że mógłbyś wciągnąć do projektu taką fajną libkę jak Apache Commons - dodaje ona sporo przydatnych metod do operacji na kolekcjach. W tym te, które w dużo bardziej czytelny sposób je łączą. Jak na przykład taki sposób łączenia list:
List<String> mergedList = CollectionUtils.collate(sortedList1, sortedList2);
https://www.baeldung.com/apache-commons-collection-utils - więcej info u baeldunga

Reply:
> The library targets older JVM releases because currently, similar functionality is provided by the Java 8's Stream API.
dlatego mam mieszane odczucia :sweat_smile: Tak czy inaczej zmieniłem kod na wykorzystywanie setów

Reply:
No w sumie streamy można łączyć w jeden. Zapomniałam o tym :P Ale i tak zaglądałabym w libkę Apache Commons, gdybym po napisaniu kodu, albo w trakcie jego pisania, stwierdziła że to co tworzę będzie zawiłe i zagmatwane, a może da się prościej i czytelniej i mało tego ktoś już to stworzył 😀